### ISSUE TITLE
Anyone interested to keep the package maintained?

### ISSUE THREAD

**Issue by mojiAh** [router5/router5#525]
Seems the package is abandoned, is there anyone who wants to collaborate maintaining the project?

**Comment by FireworksX** [comment]
I really love this routing especially after release react-router v6. I'll could collaborate with some one for improve this router.

**Comment by mrgazanfarli** [comment]
Hi! Do you still want to maintain the project? @mojiAh @FireworksX

**Comment by greydragon888** [comment]
@mrgazanfarli 

Yes! I want to maintain this project!
I haven't been able to find anything like this for many years!

**Comment by greydragon888** [comment]
@mojiAh @FireworksX @mrgazanfarli
I have some idias, how to improve this package 
Let's make router5 great again! ))

**Comment by riogod** [comment]
Ive making some refactor, tests, fixed issues from here list and add some cool features.
so you can use `@riogz/router ` (https://www.npmjs.com/package/@riogz/router). 

It fully compatable with router5 and plugin/middleware system.
PR's are welcome to improve the functionality of router.

**Comment by mrgazanfarli** [comment]
@OlIvAl did you try something?

**Comment by greydragon888** [comment]
Yes! A lot of work has already been done: dependencies have been updated and modern technologies have been introduced, type safety has been implemented. Currently, we are working on maximum code coverage with tests, identifying bugs, optimizing code, and improving documentation. I have also implemented benchmark tests.

About half of all modules are covered with tests. Unfortunately, it is not always possible to spend as much time on the project as I would like. But I plan to finish testing and documentation soon, and move on to fixing the bugs found. 

https://github.com/OlIvAl/router6/issues/52
It is not the full bug list 

As soon as I finish them, then there will be the first release

I love this library and I believe that this is currently the only solution available that can be used in a large project.

**Comment by OlIvAl** [comment]
I really love this tool and want to do a lot of work on it: test coverage, documentation, API normalization.

**Comment by greydragon888** [comment]
Hi everyone.

It's been a long time since the last update. Here's a brief status report.

The goal of refactoring and improving router5 is nearly complete, though the scope of work turned out to be significantly larger than I initially anticipated.

The refactoring work is almost done. What remains is analyzing and fixing the router.navigate and router.start methods, after which the release can be prepared. These are the most complex parts, so I'm targeting the end of January for publishing the final result.

Over the past 9 months, substantial work has been done on a deep rework of the router's architecture and infrastructure. A large number of architectural and technical issues were identified in the router5 library, and I've addressed them. The monorepo build system was redesigned from scratch, type safety was implemented, input parameter validation was added, and performance was improved—especially on performance-critical hot paths.
I also significantly improved code testing coverage and added benchmark tests.
During this work, I wrote a separate matcher library for Vitest, designed for profiling React components and hooks:
[https://github.com/greydragon888/vitest-react-profiler](https://github.com/greydragon888/vitest-react-profiler)

As a result of working on the router code, I've concluded that evolving router5 in its current form won't allow creating a truly powerful tool. I've built an extensive backlog of tasks for a full architectural refactoring (around 50 tickets) and decided not to continue maintaining router5, but rather to develop a new router based on its ideas. The current refactoring will definitely be completed and released, but further development will happen in a new project.

I'm grateful to Thomas Roch for the inspiration and intend to build a router that preserves the core ideas of router5 while being the best among all existing open source solutions.

**Comment by mojiAh** [comment]
@greydragon888 Great news, would be happy to contribute in further developments.

**Comment by greydragon888** [comment]
### A few spoilers from my backlog.

I want to create a router that is fast, simple, and secure by default.

During the analysis of the existing code, it became clear that router5 in its current form does not meet these requirements. The main limitations are architectural in nature and cannot be addressed through incremental improvements.

The functional architecture of router5 makes it difficult to ensure security and isolation. Internal methods are hard to encapsulate, and data validation is tightly coupled with business logic. This complicates the design of optimized functionality: public APIs are used internally within the router itself and are forced to re-validate already verified data.

In the new router, I will transition to a class-based architecture with a clear structural separation of the API into two layers:

* **external facade layer**, responsible for sanitization, validation, and normalization of input data;
* **internal layer**, implementing routing logic and working exclusively with guaranteed valid and normalized structures.

This separation will simplify the codebase, eliminate redundant checks, and allow optimization of hot paths without compromising security.

The route parser in router5 is one of the key performance bottlenecks. It performs an excessive number of iterations over the route graph when searching for the optimal path and requires re-sorting nodes on every configuration change. This significantly degrades navigation performance.

I plan to replace it with **rou3**. This will substantially speed up parsing and matching operations. Additionally, migration to **rou3** will simplify the parser implementation itself through more suitable data structures and algorithms.

The lack of centralized caching also negatively impacts router5 performance.

The new design will implement a unified cache manager for all routing operations. This refactoring will bring navigation complexity closer to **O(1)** for all transitions. Additionally, cache monitoring will be introduced: tracking size, memory consumption, and load, along with tools for detecting potential leaks.

The event-driven model used in router5 makes key operations - start and navigation - excessively complex and difficult to analyze. The logic branches through numerous state flags and edge cases (`isStarted`, `navigating`, `cancelled`, `completed`, etc.). This significantly complicates maintenance and testing.

I plan to replace the event-driven approach with a **finite state machine**. This will allow describing router behavior declaratively through explicit states and allowed transitions between them. The transition to FSM will also make the logic more predictable, formally verifiable, and easier to analyze and debug.

The use of callbacks in navigation additionally turns the main execution flow into hard-to-read and poorly scalable code. Also, from a DX perspective, the code looks outdated.

The new router will use an asynchronous interface oriented toward the Navigation API, with **AbortController** support. This will simplify navigation cancellation management and improve the reliability of the entire system.

**AbortController** support will also naturally simplify integration with APIs that use it. For example:

* when navigation is cancelled, related `fetch` requests can be automatically aborted;
* middleware and guards will be able to correctly respond to transition interruption;
* asynchronous operations (data loading, lazy initialization) can be guaranteed to stop without leaving "dangling" side effects.

Thus, the navigation lifecycle will be directly linked to the lifecycle of all asynchronous operations it initiates.

Additionally, I intend to clearly separate the responsibilities of plugins, middleware, and guards, as well as improve the router infrastructure overall. The goal is to simplify integration with various UI libraries and make plugin development more predictable, isolated, and secure.

Testing deserves a separate mention. I've already achieved 100% coverage and covered all known edge cases. Going forward, I plan to:

* add invariant testing;
* introduce stress tests to detect potential memory leaks;
* prepare benchmarks for basic navigation scenarios;
* use mutation testing and systematically monitor and improve the mutation score.

**The ultimate goal of my router** is not simply to rewrite router5. The goal is to create a router where performance, predictability, and security are properties of the architecture. The new router is being designed from the ground up as a system with clear boundaries of responsibility, formally described behavior, and verifiable guarantees - both at the API level and at the implementation level.

P.S. router.navigate and router.start analize and debugging completed!!! Next - release

**Comment by greydragon888** [comment]
https://github.com/greydragon888/router6/wiki - actual documentation!
Ci/cd scripts and benchmark tests of "old vs new versions" remain

**Comment by greydragon888** [comment]
Benchmark tests complete!
https://github.com/greydragon888/router6/issues/53
Should I benchmark plugins and react-components?

**Comment by greydragon888** [comment]
Release is coming!

**Comment by greydragon888** [comment]
@mojiAh @FireworksX @mrgazanfarli Need help!
https://github.com/router5/router5/issues/528

**Comment by greydragon888** [comment]
@riogod @mojiAh @FireworksX @mrgazanfarli 
The work is complete! The router is completely ready, and I've already started developing my solution! Who wants to join?
https://github.com/greydragon888/router6
https://github.com/greydragon888/real-router

**Comment by riogod** [comment]
Na, tnx! I have a self variation of router5 witch integrated in my solutions
https://www.npmjs.com/package/@riogz/router

**Comment by greydragon888** [comment]
> Na, tnx! I have a self variation of router5 witch integrated in my solutions https://www.npmjs.com/package/@riogz/router

Great job! But do you have roadmap?
What testing coverage of your package?
Do you have benchmark tests for confirmation router5 and your fork?
Do you have a full documentation of your package and new features? I could not find it

**Comment by riogod** [comment]
- ~92% coverage.
Any rewritten library requires field testing, and without it, it's impossible to assume that everything was carefully thought out during the refactoring and that all edge cases were taken into account. Synthetic benchmarks are like "a spherical horse in a vacuum"—they seem to measure something, but in reality, everything is different.

In fact, the library isn't of particular interest to the general public (except for the systems where it's used), because the generally accepted specialized solutions are no worse.

So I don't see much point, I'm just improving it for my solutions and my consumers

**Comment by greydragon888** [comment]
Excellently generated text!
Ask chat gpt, why did you not make fork of library, which isn't of particular interest 🤣🤣🤣

**Comment by riogod** [comment]
generated text? huh? 
You are so toxic. I don't wan't to continue that bullshit.
Ask your main contributor about backhanded compliment.
